# Methods

## Demographic inference

We inferred the demographic history of parental (nonadmixed) populations of *M. personata* and *M. alba* by using differential equations to generate the expected allele frequency spectrum under different evolutionary models in `moments` v.1.0.0 (Jouganous et al., 2017). As input, we used a SNP dataset that had been filtered according to GATK best practices and further trimmed to only include unlinked SNPs (r<sup>2</sup><0.1) in `scikit-allel` v.1.2.1, which estimates linkage disequilibrium using the method of Rogers and Huff (2008). We further applied `scikit-allel` to calculate the folded joint site frequency spectrum (JSFS) of *personata* and *alba*, and projected it down to an effective sample size of 16 x 16 chromosomes to account for missing data. Using `moments`, we defined three demographic models that differed in the amount and timing of gene flow between  diverging lineages: a strict isolation model ("SI"), in which *personata* and *alba* diverged in the absence of migration; an isolation-with-migration model ("IM"), in which *personata* and *alba* diverged while exchanging migrants; and a secondary contact model ("SC"), in which *personata* and *alba* diverged in isolation and later came back into contact and began exchnaging migrants. We ran 10 optimizations of parameter values using `moments` “optimize_log” method, randomly sampling starting values from a uniform distribution within the bounds of 1x10<sup>-5</sup> and 10 and allowing 5 iterations. We compared log-likelihood and AIC values across models, and and generated parameter uncertainty estimates for the best-fit model by fitting 100 bootstrapped site frequency spectra generated using the fs.sample() function on our data. To convert final parameter values to real units, we applied genome-wide mutation rate of 2.3x10 9 (Smeds et al., 2016), an effective sequence length scaled to reflect our LD-thinned SNP dataset, a generation time estimate of one year. 
