# Methods  

## Demographic inference  

We inferred the demographic history of parental (nonadmixed) populations of *M. personata* and *M. alba* by using differential equations to generate the expected allele frequency spectrum under different evolutionary models in `moments` v.1.0.0 (Jouganous et al., 2017). As input, we used a SNP dataset that had been filtered according to GATK best practices and further trimmed to only include unlinked SNPs (r<sup>2</sup><0.1) in `scikit-allel` v.1.2.1, which estimates linkage disequilibrium using the method of Rogers and Huff (2008). We further applied `scikit-allel` to calculate the folded joint site frequency spectrum (JSFS) of *personata* and *alba*, and projected it down to an effective sample size of 16 x 16 chromosomes to account for missing data. Using `moments`, we defined three demographic models that differed in the amount and timing of gene flow between  diverging lineages: a strict isolation model ("SI"), in which *personata* and *alba* diverged in the absence of migration; an isolation-with-migration model ("IM"), in which *personata* and *alba* diverged while exchanging migrants; and a secondary contact model ("SC"), in which *personata* and *alba* diverged in isolation and later came back into contact and began exchanging migrants. In all models, allowed exponential population growth in the most recent time period. We ran 10 optimizations of parameter values using `moments` “optimize” method, randomly sampling starting values from a uniform distribution within the bounds of 1x10<sup>-5</sup> and 10 and allowing a maximum of 3 iterations. We compared log-likelihood and AIC values across models, and and generated parameter uncertainty estimates for the best-fit model by fitting 100 bootstrapped site frequency spectra generated using the fs.sample() function on our data. To convert final parameter values to real units, we applied genome-wide mutation rate of 2.3x10<sup>9</sup> (Smeds et al., 2016), an effective sequence length scaled to reflect our LD-thinned SNP dataset, a generation time estimate of one year.

## Hybrid Zone Simulations  

To evaluate whether partial dominance of a plumage trait involved in mate choice might lead to displacement between phenotypic and genomic background clines, we ran forward time simulations using a modified version of the R program `HZAM` v.2.0.0 (Irwin 2020), which models on-dimensional hybrid zone dynamics in the presence of assortative mating, underdominance, and other processes. Our modification extended HZAM's base features to support partial dominance of mating trait loci, using a user-defined dominance coefficient indicating the proportion of the homozygous dominant phenotype expressed by heterozygotes. Full code and an interactive web app implementing it can be found at https://github.com/elinck/hzam_shiny. We ran 50 replicate simulations of 500 generations each with one mate choice locus and three neutral loci, a 1:2 ratio of heterospecific to homospecific matings, a dominance coefficient of 0.75, a per-species carrying capacity of 800, initial range limits of 0-0.48 and 0.52-1 (given a total transect size of 1), a population growth rate of 1.05, a density dependence effect standard deviation of 0.01, and no loss of hybrid fitness. We fit clines for mating trait and neutral loci every 5 generations using a general additive model (GAM) implemented in the R package `mgcv` v.1.8-31 (Wood 2011) with the `gam()` function, using the formula `allele_frequency ~ s(transect_location)` and the "P-ML" method. We then used the `predict_gam()` function in the R package `tidymv` v.2.2.0 to generate model predictions for evenly-spaced transect values, and used these to calculate cline width and cline center. To calculate cline width and center, we used Irwin's (2020) "eightieth percentile" approach, which defines hybrid zone width as the geographic area between the locations at which the GAM cline fit crosses *HI* (hybrid index) = 0.1 and *HI* = 0.9, and center as the region within half a dispersal distance from the location at which the cline fit crosses HI = 0.5. Using these data, we performed Wilcoxon signed rank tests to determine whether cline center differed between neutral and mating trait loci across all simulations.

# Results  

## Demographic inference  

Site frequency spectrum-based demographic inference in `moments` (**Fig. 4**) indicated a model of secondary contact was the best overall fit for the data (AIC=248644.2; log-likelihood=-124315.1), though across all initial parameter optimizations a model of strict isolation had the best median fit (AIC=415979.3; log-likelihood=-207985.6)—likely reflecting the conflict between obvious hybridization and the allopatric origins of the samples used to generate the JSFS. A model of isolation-with-migration was more poorly supported (AIC=498670.4; log-likelihood=-249329.2). Parameter estimates from the secondary contact model indicate *alba* and *personata* diverged ~77,834 years ago (SD: 2,272 years) and were isolated until ~432 years ago (SD: 528 years), at which point hybridization began. Range-wide gene flow between subspecies remains low outside the hybrid zone, with far fewer than one migrant per generation exchanged in either direction between distant parental populations in the analysis. The inferred population size of *alba* was slightly greater than that of *personata*, at ~283,484 individuals (SD: 9,806 individuals) and ~253,026 individuals (SD: 8,721 individuals), respectively.

## Hybrid Zone Simulations  

Our forward time simulations in `HZAM` suggest the partial dominance of the *alba* phenotype may be contributing to its introgression into *personata* genomic background, and the corresponding displacement of phenotypic and neutral genomic clines (**Fig 5**). After 500 generations—approximately the inferred duration of hybridization—the center of cline of the mating trait locus was significantly offset from the cline of the neutral loci in the direction of the recessive homozygous population (mean values: 0.4030753 and 0.4760450; p-value = 0.004; 95% CI: 0.02685437—0.12565345). Across all individual replicates, the mating trait locus cline center was displaced toward the recessive homozygous population 64% of the time. As expected, the mating trait locus cline was also significantly narrower than the cline of the neutral loci (mean values:  0.09505139 and 0.44965245; p-value = 0.004; 95% CI: 0.02685437—0.12565345), indicating assortative mating restricts the diffusion of mating trait alleles across the hybrid zone.
